File: python/keyczar/test/collision_test.py
#!/usr/bin/python
#
# Copyright 2008 Google Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
# 
#      http://www.apache.org/licenses/LICENSE-2.0
# 
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""
Testcases to test behavior of Keyczar Signers.

@author: arkajit.dey@gmail.com (Arkajit Dey)
"""
from __future__ import absolute_import

import os
import unittest
import datetime

from keyczar import keyczar
from keyczar import util

class CollisionTest(unittest.TestCase):

  TEST_DATA = os.path.realpath(os.path.join(os.getcwd(), "test-data", "special-case"))

  def setUp(self):
    self.input = "This is some test data"

  def __testDecrypt(self, subdir):
    path = os.path.join(self.TEST_DATA, subdir)
    crypter = keyczar.Crypter.Read(path)
    active_ciphertext = util.ReadFile(os.path.join(path, "1.out"))
    primary_ciphertext = util.ReadFile(os.path.join(path, "2.out"))
    active_decrypted = crypter.Decrypt(active_ciphertext)
    self.assertEqual(self.input, active_decrypted)
    primary_decrypted = crypter.Decrypt(primary_ciphertext)
    self.assertEqual(self.input, primary_decrypted)

  def __readGoldenOutput(self, subdir, public=False):
    path = os.path.join(self.TEST_DATA, subdir)
    if not public:
      czar = keyczar.Verifier.Read(path)
    else:
      czar = keyczar.Verifier.Read(os.path.join(self.TEST_DATA, 
        subdir+".public"))
    active_sig = util.ReadFile(os.path.join(path, "1.out"))
    primary_sig = util.ReadFile(os.path.join(path, "2.out"))
    return (czar, active_sig, primary_sig)

  def __testVerify(self, subdir):
    (verifier, active_sig, primary_sig) = self.__readGoldenOutput(subdir)
    self.assertTrue(verifier.Verify(self.input, active_sig))
    self.assertTrue(verifier.Verify(self.input, primary_sig))
 
  def __testVerifyAttached(self, subdir, secret="", public =False):
    path = os.path.join(self.TEST_DATA, subdir)
    verifypath = path
    if public:
      verifypath = path +".public"
    ext = ".attached"
    if secret:
      ext = "." + secret + ext
    message = util.ReadFile(os.path.join(path, "2" + ext))
    verifier = keyczar.Verifier.Read(verifypath)
    self.assertTrue(verifier.AttachedVerify(message, secret))

  def __testVerifyTimeout(self, subdir, expired =False, public =False):
    path = os.path.join(self.TEST_DATA, subdir)
    verifypath = path
    if public:
      verifypath = path +".public"
    date = lambda: datetime.datetime(2012, 12, 21, 11, 6)
    if expired:
      date = lambda:datetime.datetime(2012, 12, 21, 11, 16)
    sig = util.ReadFile(os.path.join(path, "2.timeout"))
    verifier = keyczar.TimeoutVerifier.Read(verifypath)
    verifier.SetCurrentTimeFunc(date)
    self.assertEqual(verifier.Verify(self.input, sig), not expired)

  def testAesKeyDecrypt(self):
    self.__testDecrypt(os.path.join("key-collision", "aes"))

  def testRsaKeyDecrypt(self):
    self.__testDecrypt(os.path.join("key-collision", "rsa"))

  def testHmacKeyVerify(self):
    self.__testVerify(os.path.join("key-collision", "hmac"))

  def testHmacKeyVerifyTimeoutSuccess(self):
    self.__testVerifyTimeout(os.path.join("key-collision", "hmac"))

  def testHmacKeyVerifyTimeoutExpired(self):
    self.__testVerifyTimeout(os.path.join("key-collision", "hmac"), True)

  def testHmacKeyVerifyAttached(self):
    self.__testVerifyAttached(os.path.join("key-collision", "hmac"))

  def testHmacKeyVerifyAttachedSecret(self):
    self.__testVerifyAttached(os.path.join("key-collision", "hmac"), "secret")

  def testDsaKeyVerify(self):
    self.__testVerify(os.path.join("key-collision", "dsa"))

  def testDsaKeyVerifyTimeoutSuccess(self):
    self.__testVerifyTimeout(os.path.join("key-collision", "dsa"))

  def testDsaKeyVerifyTimeoutExpired(self):
    self.__testVerifyTimeout(os.path.join("key-collision", "dsa"), True)

  def testDsaKeyVerifyAttached(self):
    self.__testVerifyAttached(os.path.join("key-collision", "dsa"))

  def testDsaKeyVerifyAttachedSecret(self):
    self.__testVerifyAttached(os.path.join("key-collision", "dsa"), "secret")

  def testRsaKeyVerify(self):
    self.__testVerify(os.path.join("key-collision", "rsa-sign"))

  def testRsaKeyVerifyTimeoutSuccess(self):
    self.__testVerifyTimeout(os.path.join("key-collision", "rsa-sign"))

  def testRsaKeyVerifyTimeoutExpired(self):
    self.__testVerifyTimeout(os.path.join("key-collision", "rsa-sign"), True)

  def testRsaKeyVerifyAttached(self):
    self.__testVerifyAttached(os.path.join("key-collision", "rsa-sign"))

  def testRsaKeyVerifyAttachedSecret(self):
    self.__testVerifyAttached(os.path.join("key-collision", "rsa-sign"), 
      "secret")

  def tearDown(self):
    self.input = None

def suite():
  suite = unittest.TestSuite()
  suite.addTests(unittest.TestLoader().loadTestsFromTestCase(CollisionTest))
  return suite

if __name__ == "__main__":
  unittest.main(defaultTest='suite')

File: java/code/tests/org/keyczar/CollisionTest.java
package org.keyczar;

import java.io.RandomAccessFile;

import org.apache.log4j.Logger;
import org.junit.Test;
import org.keyczar.interfaces.KeyczarReader;
import org.keyczar.util.Base64Coder;
import org.keyczar.util.Clock;

import junit.framework.TestCase;


public class CollisionTest extends TestCase {
    private static final Logger LOG = Logger.getLogger(CollisionTest.class);
    private static final String TEST_DATA = "./testdata/key-collision";
    private String input = "This is some test data";
    
    private final void testDecrypt(String subDir) throws Exception {
        testDecrypt(new KeyczarFileReader(testData(subDir)), subDir);
    }
    
    private String testData(String subDir){
      return TEST_DATA + subDir;
    }

    private final void testDecrypt(KeyczarReader reader, String subDir)
        throws Exception {
      Crypter crypter = new Crypter(reader);
      RandomAccessFile activeInput =
        new RandomAccessFile(testData(subDir) + "/1.out", "r");
      String activeCiphertext = activeInput.readLine(); 
      activeInput.close();
      RandomAccessFile primaryInput =
        new RandomAccessFile(testData(subDir) + "/2.out", "r");
      String primaryCiphertext = primaryInput.readLine();
      primaryInput.close();
      String activeDecrypted = crypter.decrypt(activeCiphertext);
      assertEquals(input, activeDecrypted);
      String primaryDecrypted = crypter.decrypt(primaryCiphertext);
      assertEquals(input, primaryDecrypted);
    }
    
    private final void testVerify(String subDir) throws Exception {
        Verifier verifier = new Verifier(testData(subDir));
        RandomAccessFile activeInput =
          new RandomAccessFile(testData(subDir) + "/1.out", "r");
        String activeSignature = activeInput.readLine(); 
        activeInput.close();
        RandomAccessFile primaryInput =
          new RandomAccessFile(testData(subDir) + "/2.out", "r");
        String primarySignature = primaryInput.readLine();
        primaryInput.close();

        assertTrue(verifier.verify(input, activeSignature));
        assertTrue(verifier.verify(input, primarySignature));
     }
    
    private final void testVerifyAttached(String subDir,String hidden) throws Exception {
      Verifier verifier = new Verifier(testData(subDir));
      String hiddenExt="";
      if(hidden != "")
        hiddenExt = "." + hidden;
        RandomAccessFile activeInput =
          new RandomAccessFile(testData(subDir) + "/1"+hiddenExt+".attached", "r");
        String activeSignature = activeInput.readLine(); 
        activeInput.close();
        
        RandomAccessFile primaryInput =
              new RandomAccessFile(testData(subDir) + "/2"+hiddenExt+".attached", "r");
            String priarySignature = primaryInput.readLine(); 
            primaryInput.close();
        
        assertTrue(verifier.attachedVerify(Base64Coder.decodeWebSafe(activeSignature), 
            hidden.getBytes(Keyczar.DEFAULT_ENCODING)));
        
        assertTrue(verifier.attachedVerify(Base64Coder.decodeWebSafe(priarySignature), 
            hidden.getBytes(Keyczar.DEFAULT_ENCODING)));
    }
    
    public class EarlyClock implements Clock{
      public long now(){
        return 1356087960000L;
      }
    }
    
    public class LateClock implements Clock{
      public long now(){
        return 1356088560000L;
      }
    }
    
    private final void testTimeoutVerifier(String subDir) throws Exception {
      TimeoutVerifier verifier = new TimeoutVerifier(testData(subDir));
      verifier.setClock(new EarlyClock());
        RandomAccessFile activeInput =
          new RandomAccessFile(testData(subDir) + "/2.timeout", "r");
        String activeSignature = activeInput.readLine(); 
        activeInput.close();
        
        RandomAccessFile primaryInput =
              new RandomAccessFile(testData(subDir) + "/2.timeout", "r");
            String primarySignature = primaryInput.readLine(); 
            primaryInput.close();
    
        assertTrue(verifier.verify(input, activeSignature));
        assertTrue(verifier.verify(input, primarySignature));
    }
    
      private final void testTimeoutVerifierExpired(String subDir) throws Exception {
        TimeoutVerifier verifier = new TimeoutVerifier(testData(subDir));
      verifier.setClock(new LateClock());
        RandomAccessFile activeInput =
          new RandomAccessFile(testData(subDir) + "/2.timeout", "r");
        String activeSignature = activeInput.readLine(); 
        activeInput.close();
        
        RandomAccessFile primaryInput =
        new RandomAccessFile(testData(subDir) + "/2.timeout", "r");
      String primarySignature = primaryInput.readLine(); 
        primaryInput.close();
        
        assertFalse(verifier.verify(input, activeSignature)); 
        assertFalse(verifier.verify(input, primarySignature));
    }
      
    @Test 
    public final void testHmacVerify() throws Exception {
      testVerify("/hmac");
    }
    
    @Test 
    public final void testHmacVerifyAttached() throws Exception {
      testVerifyAttached("/hmac", "");
    }
    
    @Test 
    public final void testHmacVerifyAttachedSecret() throws Exception {
      testVerifyAttached("/hmac", "secret");
    }
    
    @Test 
    public final void testHmacVerifyTimeoutSuccess() throws Exception {
      testTimeoutVerifier("/hmac");
    }
    
    @Test 
    public final void testHmacVerifyTimeoutExpired() throws Exception {
      testTimeoutVerifierExpired("/hmac");
    }
    
    
    @Test 
    public final void testDsaVerify() throws Exception {
      testVerify("/dsa");
    }
    
    @Test 
    public final void testDsaVerifyAttached() throws Exception {
      testVerifyAttached("/dsa", "");
    }
    
    @Test 
    public final void tesDsaVerifyAttachedSecret() throws Exception {
      testVerifyAttached("/dsa", "secret");
    }
    
    @Test 
    public final void testDsaVerifyTimeoutSuccess() throws Exception {
      testTimeoutVerifier("/dsa");
    }
    
    @Test 
    public final void testDsaVerifyTimeoutExpired() throws Exception {
      testTimeoutVerifierExpired("/dsa");
    }
    
    @Test 
    public final void testRsaVerify() throws Exception {
      testVerify("/rsa-sign");
    }
    
    @Test 
    public final void testRsaVerifyAttached() throws Exception {
      testVerifyAttached("/rsa-sign", "");
    }
    
    @Test 
    public final void tesRsaVerifyAttachedSecret() throws Exception {
      testVerifyAttached("/rsa-sign", "secret");
    }
    
    @Test 
    public final void testRsaVerifyTimeoutSuccess() throws Exception {
      testTimeoutVerifier("/rsa-sign");
    }
    
    @Test 
    public final void testRsaVerifyTimeoutExpired() throws Exception {
      testTimeoutVerifierExpired("/rsa-sign");
    }
    
    @Test
    public final void testAesDecrypt() throws Exception {
      testDecrypt("/aes");
    }
    
    @Test
    public final void testRsaDecrypt() throws Exception {
      testDecrypt("/rsa");
    }
    
    
}
